add tests for import map composition and a "cascading" composition implementation

michaelficarra opened the pull request:
In the hopes that an implementation will help make the case for #137, I present an implementation of cascading import map composition and tests that demonstrate its behaviour.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/michaelficarra/import-maps/pull/167.html" title="Last updated on Mar 5, 2021, 9:39 PM UTC (22f5238)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/import-maps/167/3467c3c...michaelficarra:22f5238.html" title="Last updated on Mar 5, 2021, 9:39 PM UTC (22f5238)">Diff</a>

michaelficarra committed:
add tests for composition and a "cascading" composition implementation

michaelficarra committed:
fix linting (though the code is arguably worse now)

michaelficarra committed:
switch maps from URLs to strings

domenic commented:
I've discussed this with some folks offline and we'd be willing to go this route if it doesn't add too much complexity.

However, to get a sense of that complexity, we'd need a more complete update to the reference implementation and tests. In particular, this only creates a merging procedure, but we also need to see and test how the resolution works, and how much complexity is added there. And we need to update the types and terminology used; right now the reference implementation represents addresses using `URL` objects, but if we're changing the right-hand side from a URL to a module specifier, then we need to stop using `URL` objects there.

In particular, in offline discussions you've explained how the mental model for web developers is something like

```js
// Turns relative URL-like specifiers into absolute-URL-like ones
let specifier = normalize(specifierFromSourceText);

specifier = apply(importMapB, specifier);
specifier = apply(importMapA, specifier);

// Throws on specifiers that are not absolute-URL-like. Identity transform otherwise.
let url = toURL(specifier);
```

We need to see this manifested in the reference implementation, where it will probably be something more like

```js
// Inside resolve() in resolver.js

// ... existing code to get normalizedSpecifier ...
const mappedSpecifier = apply(mergedImportMap, normalizedSpecifier, scriptURL);

const url = tryURLParse(mappedSpecifier);

if (!url) {
  throw new TypeError(`Unmapped bare specifier "${specifier}"`);
}

// ... existing code to reject un-implemented built-in modules ...

return url;
```

michaelficarra commented:
@domenic Take a look at the updated changes. I've updated/added tests and the resolving implementation.

bakkot committed:
deduplicate logic between resolver and composer

bakkot committed:
empty commit to rerun CI

domenic commented on reference-implementation/__tests__/composition.js:
```diff
@@ -0,0 +1,367 @@
+
+describe('Composition', () => {
+  it('should compose with the empty map on the left', () => {
+    const map = parseFromString(JSON.stringify({
```
I'd like to continue representing maps in the tests as strings.

domenic commented on reference-implementation/__tests__/composition.js:
```diff
@@ -0,0 +1,367 @@
+
+const mapBaseURL = new URL('https://example.com/app/index.html');
+
+function composeMaps(mapLikes) {
```
It'd be nice to add a comment explaining that this functions serves a few purposes (as far as I can tell):

- Allowing the input to be plain JS objects, instead of import map objects or strings
- Automatically assumes the input will use mapBaseURL.
- Allowing writing tests that list all inputs, instead of creating the first, then appending xN

domenic commented on reference-implementation/__tests__/composition.js:
```diff
@@ -0,0 +1,367 @@
+    const newMap = parseFromString(JSON.stringify(mapLike), mapBaseURL);
+    map = appendMap(map, newMap);
+  }
+  return JSON.parse(JSON.stringify(map));
```
What is the purpose of this? It seems like it's going to hide something from us...

michaelficarra replied:
It was left over from when we had URL objects in the result, as a way to make it easier to write tests. Removed now.

domenic commented on reference-implementation/__tests__/composition.js:
```diff
@@ -0,0 +1,367 @@
+    });
+  });
+
+  it('should compose maps that are using the virtualisation patterns we expect to see in the wild', () => {
```
```suggestion
  it('should compose maps that are using the virtualization patterns we expect to see in the wild', () => {
```

domenic commented on reference-implementation/__tests__/composition.js:
```diff
@@ -0,0 +1,367 @@
+    });
+  });
+
+  it('should compose maps that interact via cascading', () => {
```
Can you add some composition tests with relative URLs, specifically ones that are different representations of each other? For example, both /b and ../b refer to the same URL, and so if one map uses the former and another the latter, you should still get this kind of cascading behavior.

This should probably be done for the next several tests, at least until you feel confident we've covered the potential places that this could show up.

domenic commented on reference-implementation/__tests__/composition.js:
```diff
@@ -0,0 +1,367 @@
+    });
+  });
+
+  it('should compose "nested" scopes', () => {
```
This seems to make a decision on https://github.com/WICG/import-maps/issues/153. Which I'm fine with; that issue is unresolved just because I've been waiting for someone to chime in, and you've done it. But can we add a simple test, like the one from (the scopes part of) https://github.com/WICG/import-maps/issues/153#issuecomment-513843388 , before getting in to cascading?

domenic commented on reference-implementation/__tests__/composition.js:
```diff
@@ -0,0 +1,367 @@
+      {
+        imports: {
+          'utils/': 'moment/lib/utils/',
+          'is-date': 'moment/lib/utils/is-date.js'
```
Can you test that this *doesn't* work within a single map? For example an entry

```js
'foo': 'utils/foo.js'
```

should result in... just a bare specifier in the output, I think?

bakkot replied:
> should result in... just a bare specifier in the output, I think?

Yes, except that [this](https://github.com/WICG/import-maps/pull/167#discussion_r314064807) will change "the output is just a bare specifier" to the equivalent "the output is an empty array" (and a warning).

domenic commented on reference-implementation/__tests__/composition.js:
```diff
@@ -0,0 +1,367 @@
+    });
+  });
+
+  it('should url normalize things which have composed into URLs', () => {
```
```suggestion
  it('should URL-normalize things which have composed into URLs', () => {
```

domenic commented on reference-implementation/__tests__/parsing-addresses.js:
```diff
@@ -49,14 +49,14 @@ describe('Relative URL-like addresses', () => {
   });
 
-  it('should ignore percent-encoded variants of ./, ../, or /', () => {
+  it('should treat percent-encoded variants of ./, ../, or / as non-URLs', () => {
```
So now we have bare specifiers in the output. This is not great for the ultimate, composed import map; ideally we'd give people a warning in such cases. I.e. it'd be ideal if bare specifiers in the first import map were discarded with a warning, bare specifiers in the second that did not map to anything from the first were discarded with a warning, etc.

I'm not sure whether that's feasible with the currently-envisioned architecture. (This is part of why I anticipated this being an even bigger change than it currently is.) I don't necessarily think we should demand it, since it's just about good warnings at parse time (instead of errors at use time), but we should weigh it.

Of course, if there's some nice way to make that work with minimal disruption, that'd be great as well.

bakkot replied:
We should be able to give warnings for this and remove the specifiers from the output, but it would happen at composition time, not parse time. (You can conceptualize this as starting the document with an empty map and composing any that are present, in order.)

bakkot replied:
To spell this out a little more:

You can't give a warning for a bare specifier on the right-hand side _at parse time_, because that can be legal when adding an import map to a page which provides a mapping for that specifier. For example,

```html
<script type="importmap">
{
  "imports": {
    "moment": "/node_modules/moment/src/moment.js"
  }
}
</script>
<script type="importmap">
{
  "imports": {
    "m": "moment"
  }
}
</script>
```
is fine, despite the second map having the bare specifier `"moment"` on the RHS, because that `"moment"` is mapped to a real URL by the first map.

But you _can_ give a warning for a bare specifier on the RHS which is still present _after composition_, since maps only cascade to earlier maps, not later ones. And you can use this mechanism to give warnings for the first map on the page by acting as if every page starts out having the empty import map `{}` loaded already. For example,

```html
<script type="importmap">
{
  "imports": {
    "moment": "/node_modules/moment/src/moment.js",
    "lodash": "lodash/index.js"
  }
}
</script>
<script type="importmap">
{
  "imports": {
    "jquery": "javascript:bad",
    "l": "lodash"
  }
}
</script>
```

can give a warning for `"lodash"` when adding the first map and for both `"jquery"` and `"l"` when adding the second map, since you know when each map is added to the page that those point to bare specifiers. The PR has been updated with this behavior.

domenic commented on reference-implementation/__tests__/parsing-addresses.js:
```diff
@@ -152,25 +144,20 @@ describe('Absolute URL addresses', () => {
+        http: ['http://good/'],
+        https: ['https://good/'],
+        ftp: ['ftp://good/'],
+        import: ['import:bad'],
```
The loss of these warnings is also troubling. These aren't even bare specifiers... they're invalid absolute URL specifiers. Why did we lose the ability to reject them?

bakkot replied:
Absolute URL specifiers with non-fetch schemes are treated as non-URLs, so in this case the specifier `import:bad` would be successfully *parsed* but not allowed to be the result of resolution. In [a separate test](https://github.com/WICG/import-maps/blob/ccaf0c062738a2c1b93470ffffa297ed3f251b33/reference-implementation/__tests__/resolving.js#L94-L108), we show how non-fetch-scheme-URL module specifiers like this one can be made to map to resolvable URLs without error.

After addressing [your earlier comment about stripping bare specifiers](https://github.com/WICG/import-maps/pull/167/files#r314491076), these would end up removed with a warning at composition time.

domenic commented on reference-implementation/__tests__/resolving-builtins.js:
```diff
@@ -155,4 +168,16 @@ describe('Fallbacks with built-in module addresses', () => {
+    expect(resolveUnderTest('twoGoodBuiltins')).toMatchURL(BLANK);
+  });
+
+  it('should fail for [built-in, fetch scheme, fetch scheme]', () => {
```
These aren't quite correct. resolve() would return the first URL which was successfully fetched, most likely. The complexities there (e.g. making resolve() async or network dependent) are why we've kept this as unimplemented. It'd be better to stick to only supporting the [built-in, fetch scheme URL] pattern for now in the fallbacks.

bakkot replied:
For simplicity, this implementation assumes that all web resource fetches succeed (which seems to match the current tested behavior). We can artificially restrict the resolution algorithm to only accept fallback lists supported by the current implementation. Or we could use some different strategy to determine a success status.

domenic commented on reference-implementation/__tests__/resolving.js:
```diff
@@ -63,6 +60,53 @@ describe('Unmapped', () => {
+describe('the empty string', () => {
+  it('should fail for an unmapped empty string', () => {
+    const resolveUnderTest = makeResolveUnderTest(`{}`);
+    console.warn = () => {};
```
Why is this necessary? Same below. It seems we should not get a warning if we're also getting a TypeError... But even if we do, we should test for the presence of the warning.

michaelficarra replied:
The warning comes from parsing and the thrown `TypeError` comes from resolution. They're composed in `makeResolveUnderTest`. I'll assert that the warnings are present in the same way we do in `testWarningHandler`.

domenic commented on reference-implementation/lib/resolver.js:
```diff
@@ -1,100 +1,60 @@
-  const topLevelImportsMatch = resolveImportsMatch(normalizedSpecifier, parsedImportMap.imports);
-  if (topLevelImportsMatch !== null) {
-    return topLevelImportsMatch;
+exports.resolve = (specifier, parsedImportMap, scriptURLparameter) => {
```
If it's easy it would be nice to have this take a `URL` as its third argument, directly. (Bad JavaScript API design, but good for a reference implementation mirroring the spec.) I kind of thought it did previously...

domenic commented on reference-implementation/lib/utils.js:
```diff
@@ -19,23 +19,56 @@ exports.tryURLParse = (string, baseURL) => {
 exports.tryURLLikeSpecifierParse = (specifier, baseURL) => {
+  if (specifier === '') {
+    console.warn(`Invalid empty string specifier.`);
+    return { type: 'invalid' };
```
Maybe it'd be better to communicate the warning as a field in the return type, which can then be turned into a thrown TypeError, or console.warn()ed, by the caller? (Related to above comment about tests that set console.warn to a no-op.)

domenic commented on reference-implementation/__tests__/parsing-addresses.js:
```diff
@@ -334,17 +316,17 @@ describe('Failing addresses: mismatched trailing slashes', () => {
 describe('Other invalid addresses', () => {
-  it('should ignore unprefixed strings that are not absolute URLs', () => {
+  it('should treat unprefixed strings that are not absolute URLs as non-URLs', () => {
     for (const bad of ['bar', '\\bar', '~bar', '#bar', '?bar']) {
```
"bad" is probably no longer the right variable name here

domenic commented on reference-implementation/lib/resolver.js:
```diff
@@ -1,100 +1,60 @@
-      } else if (addresses.length === 2 &&
-                 addresses[0].protocol === BUILT_IN_MODULE_PROTOCOL &&
-                 addresses[1].protocol !== BUILT_IN_MODULE_PROTOCOL) {
+exports.getFallbacks = (normalizedSpecifier, parsedImportMap, resolutionContext) => {
```
resolutionContext -> scriptURL, maybe?

bakkot replied:
This is used in [the composer](https://github.com/WICG/import-maps/blob/ccaf0c062738a2c1b93470ffffa297ed3f251b33/reference-implementation/lib/composer.js#L23) to resolve fallbacks for a scope in the context of the existing map based on that scope (i.e., with the scope as the resolution context). In that case there's not yet a "script"; hence the name of this parameter.

domenic reviewed (commented):
Thanks so much for the work on this, and sorry for the delay. This is a big change, but you've clearly gone out of your way to create a battery of tests, as well as do minor things like match the coding style. That makes it much easier.

I'm also heartened by the relative decrease in code size in resolver.js. Some of that may be incidental (e.g. facotring out some of the logic into tryURLLikeSpecifierParse, or removing our safeguards against some not-implemented cases, or using a functional style instead of a (more spec-translation-friendly) imperative. But the getFallbacks algorithm seems like an actual win.

The main changes I'd like to see is adding a few more of the tests in locations I mentioned inline. Also, I'd like @hiroshige-g to take a look at the new/modified tests, and see if he has any of his own suggestions for devious tests to add. For example, it seems possible to me that there aren't enough tests exploring the consequences of mapping bare specifiers twice, and how that might interact with scopes; I see good coverage for absolute URLs but not as much for bare specifiers. Also, I'd like to make sure we have coverage for nailing down an answer to #153.

Overall, it seems likely this can work. I continue to maintain that the result is a more complex mental model than the current specifiers -> URLs setup; I know you disagree. But I'm willing to accept that this increase in complexity in exchange for the compositional benefits you've worked to explain.

Finally, @hiroshige-g should also take a look at the new code logic, and judge if there are any problematic aspects. In particular I want to be sure that the new resolve() function is not too expensive in some way, since that is semi-hot. It seems to have the same algorithmic complexity over various inputs, but I haven't analyzed it rigorously, and it may benefit from some optimizations (e.g. there may be places you could reduce the number of O(n) iterations over a list).

michaelficarra committed:
Update reference-implementation/__tests__/composition.js

Co-Authored-By: Domenic Denicola <d@domenic.me>

michaelficarra committed:
Update reference-implementation/__tests__/composition.js

Co-Authored-By: Domenic Denicola <d@domenic.me>

guybedford reviewed (commented):
I haven't followed the exact details but the overall approach seems like a sensible model, and while adding complexity it doesn't seem to cross any lines to me.

I just thought I would mention that my main concern here would be that with this proposal, once a URL has been overridden, the original can never be referenced again.

So later import maps effectively have less control than earlier ones. Maybe that's a bug or a feature though! (interestingly it can actually be seen as a security feature to restrict builtins...).

bakkot commented:
@guybedford That's intentional. Otherwise the README's example of [denying access to a built-in module](https://github.com/WICG/import-maps/#denying-access-to-a-built-in-module) would be impossible; see #137 for the motivations behind this PR.

michaelficarra committed:
remove JSON round-trip

michaelficarra committed:
inline JSON representation of input import maps in composition tests

michaelficarra committed:
add comment and parameter default for composition test helper function

michaelficarra committed:
different representations of URLs

michaelficarra committed:
move console warning out of tryURLLikeSpecifierParse

michaelficarra committed:
assert warnings

michaelficarra committed:
assume resolve's scriptURL parameter is a URL

michaelficarra committed:
add another test for maps not cascading to themselves

michaelficarra committed:
add another test for merge-within-scopes strategy (taken from #153)

michaelficarra committed:
composition warns if there's a non-URL in the RHS of the result

michaelficarra committed:
add another test for bare specifier stripping

michaelficarra commented:
@domenic All comments addressed. I've left the commits separate to make it easier to review only the changes.

michaelficarra committed:
bad -> nonURL

michaelficarra commented:
Ping @domenic. No rush, but I wanted to make sure you didn't miss that this is ready for review. Also, @hiroshige-g I believe we are waiting on a review from you as well.

domenic committed:
Change duplicate test names

domenic committed:
Always use std:blank and variable prefix

This gives us a slightly better chance of porting these to web platform tests in the future...

domenic commented:
Thanks, this looks good. I have run out of time today, but I will work on porting the reference implementation updates to the spec text next week.

domenic commented:
I was trying to explain to @hiroshige-g the compositional mental model you explained to me offline. In particular I think your model was something like: authors expect the import maps to apply in reverse (?) order, eventually bottoming out at a "basic specifier", which gets an identity transform from specifier string -> URL space. But when trying to reproduce this on [the "virtualization patterns we expect to see in the wild"](https://github.com/WICG/import-maps/pull/167/files#diff-bc8e235e65b9e141f5db989de8f790d9R133) example, it didn't quite work that way.

In particular, consider a module that does `import "std:blank"` from inside `https://built-in-enhancement-2/`.

```js
const specifierFromSourceText = "std:blank";

const specifier = normalizeToURLString(specifierFromSourceText);
// -> "std:blank"

specifier = apply(map3, specifier);
// -> https://built-in-enhancement-3/

specifier = apply(map2, specifier);
// -> no change...

specifier = apply(map1, specifier);
// -> no change...

const url = parseToURL(specifier);
// -> https://built-in-enhancement-3/
```

whereas the test states the result should be `https://built-in-enhancement-1/`.

If the intuition is non-reversed, that fixes the result when the referring script's URL is inside `https://built-in-enhancement-2/`, but then we have a problem when working from inside `https://built-in-enhancement-1/`:


```js
const specifierFromSourceText = "std:blank";

const specifier = normalizeToURLString(specifierFromSourceText);
// -> "std:blank"

specifier = apply(map1, specifier);
// -> std:blank

specifier = apply(map2, specifier);
// -> https://built-in-enhancement-2/

specifier = apply(map3, specifier);
// -> no change...

const url = parseToURL(specifier);
// -> https://built-in-enhancement-2/
```

whereas the test says it should be `std:blank`.

I know we've implemented a different version of this (which I prefer), which involves creating a single composed import map. We should keep doing that. But I want to try to recapture the intuition you were explaining to me about how authors might think about it.

bakkot commented:
The simple mental model we explained makes sense only when all of the mappings are similarly specific. So, if all mappings are in the same scope or all at the top level, that *is* how composition works. It gets more complicated when adding an import map with less-specific scopes on top of an import map with more-specific scopes: in that case you would not want to apply a mapping from a less-specific scope when a more-specific scope gives a mapping for the module specifier of interest, since more specific scopes are considered to be higher precedence.

(At the time of our conversations with you, we hadn't yet considered the interactions of scope specificity with composition.)

domenic committed:
Checkpoint: spec concatenating

domenic committed:
Checkpoint: spec parsing a specifier

domenic committed:
Checkpoint: specify "get the fallbacks"

domenic committed:
Spec new "resolve"

domenic committed:
Update parser spec

domenic committed:
Properly make the URL types flow through

domenic committed:
100% coverage

domenic committed:
Move composing spec section after parsing

domenic committed:
More rearranging

domenic commented:
OK, the spec updates are almost done. (And the corresponding reference implementation tweaks.) What's remaining:

- [ ] Wait for https://github.com/whatwg/infra/pull/265 to get merged and reference that (currently there's a placeholder
- [ ] Add an example or two or three of concatenating import maps, and explain the mental model. This replaces the [previous, now-inaccurate example](https://wicg.github.io/import-maps/#example-merging). I would appreciate if @michaelficarra or @bakkot were willing to take that on (pushing to this branch).

hiroshige-g commented:
Thanks for clarification. So provided the scopes of different import maps don't interfere, then we can interpret the cascading as "apply import maps in-order, while scopes are exempted from further overriding by later import maps".

For example in https://github.com/WICG/import-maps/issues/137#issuecomment-512951464, the `"/kvs-v2-polyfill.mjs"` scope in the first import map is exempted from overriding by the second import map.

hiroshige-g commented:
I'm still highly concerned about the complexity (to implementation and to users) added by this.
Especially the change adds mixture of URL and non-URLs and parsing/serializing/reparsing URLs, which makes it harder to understand the import maps behavior.

I'd like to defer the final decision at least after we pass Chromium-internal reviews (preparing...) on the basic parts of import maps that are already reflected in the current draft spec.

bakkot commented:
@domenic 

> Add an example or two or three of concatenating import maps, and explain the mental model. This replaces the previous, now-inaccurate example. I would appreciate if @michaelficarra or @bakkot were willing to take that on (pushing to this branch).

Will do. I'll try to get to this early next week.

@hiroshige-g

> So provided the scopes of different import maps don't interfere, then we can interpret the cascading as "apply import maps in-order, while scopes are exempted from further overriding by later import maps".

In reverse order, but yes.

An (equivalent) way of describing it is "find the most specific applicable scope which provides a mapping for the module specifier being evaluated, across all maps, and apply maps in reverse order starting from the map with that scope".

Or, if you are thinking about about composition directly, you can (again equivalently) say "when merging in a new map, the keys on its RHS are evaluated in the context of the existing map, just as any other module specifier would be. Scoped mappings are evaluated as if in that scope."

hiroshige-g commented:
Another example of breaking the "apply import maps in-order" semantics:
In the current PR, IIUC

```
<script type="importmap">
{
    "imports": {
      "bar/": "https://bar.example.com/",
      "bar/baz/": "https://baz.example.com/"
     }
}
</script>
<script type="importmap">
{
    "imports": { "foo/": "bar/" }
}
</script>
```

is combined into

```
<script type="importmap">
{
    "imports": { "foo/": "https://bar.example.com/" }
}
</script>
```

and thus `foo/baz/a.js` is mapped into `https://bar.example.com/baz.a.js`,
not `foo/baz/a.js` -> `bar/baz/a.js` -> `https://baz.example.com/a.js`.

This is inevitable though...

hiroshige-g commented:
Er, can this PR already result in exponential-sized output?
For example:

```
<script type="importmap">
{
    "imports": {
      "a1": "https://example.com/a1",
      "a2": "https://example.com/a2"
     }
}
</script>
<script type="importmap">
{
    "imports": {
      "b1": ["a1", "a2"],
      "b2": ["a1", "a2"]
     }
}
</script>
<script type="importmap">
{
    "imports": {
      "c1": ["b1", "b2"],
      "c2": ["b1", "b2"]
     }
}
...
</script>
```

result in `O(2^n)` addresses for `n` import maps?

bakkot commented:
> Er, can this PR already result in exponential-sized output?

Hm. As implemented, yes, because there's nothing forbidding having the same specifier appear multiple times in a fallback list (even within a single map). It doesn't really make sense to allow that, though. If fallback lists are deduplicated during composition I believe that problem goes away.

hiroshige-g commented:
I agree that there are no valid use cases for exponentially-exploding cases. But even with deduplication, trailing-slash matching can result in exponential-sized output.

```
<script type="importmap">
{
    "imports": {
      "a1/": "https://example.com/1/",
      "a2/": "https://example.com/2/"
     }
}
</script>
<script type="importmap">
{
    "imports": {
      "b1/": ["a1/1/", "a2/1/"],
      "b2/": ["a1/2/", "a2/2/"]
     }
}
</script>
<script type="importmap">
{
    "imports": {
      "c1": ["b1", "b2"],
      "c2": ["b1", "b2"]
     }
}
...
</script>
```

bakkot commented:
Ah, yes, I keep forgetting about package-prefix mappings.

In this case that seems to be what the user is asking for, though, so I'm not sure to what extent it should be regarded as a problem. In an actual implementation it could be represented in memory without the exponential overhead, if that's the worry.

On the other hand, would browsers in practice cap how many times they attempted to fetch a given module? If so, it seems reasonable to cap the number of fallbacks to that number.

guybedford commented:
Might this be an argument for treating cascade resolution as actually performing the fallback selection?

lemanschik commented:
just my 5 cent doesen't that overall add a big extra problem for that we need to create additional tooling i see we shifting the problems around we change the place where they happen this should when get done be done only inside a serviceWorker that that it gives a JS static analyzeable way to audit track what does resolve to what.

domenic commented:
We went a different direction for multiple import maps support.
